### PR TITLE
Bootloader updates for fmu-v2

### PR DESCRIPTION
**Pull request by bkueng**
Since we don't seem to have enough bandwidth on the QGC side for https://github.com/mavlink/qgroundcontrol/issues/6488, this proposes an alternative way to update the Bootloader on v2 boards that does not rely on QGC changes.

### How does it work?
The BL binary is included in the Firmware binary, and the user can then activate the update through a parameter that gets checked on the next boot.

### Steps for a user
- Insert an SD card (for the bootlog)
- Flash master of fmu-v2 target (e.g. through QGC)
- Enable `SYS_BL_UPDATE`
- Reboot & Wait
- Flash fmu-v3 target

I tested it around 10 times, works w/o problems, and the actual BL update takes only about a second.

Drawback: increased Flash usage due to the inclusion of the BL binary.

Fixes: https://github.com/PX4/Firmware/issues/9562

### CODE REVIEW

**Comment by dagar** [comment]
We could expand this in the future to build and include the latest stable bootloader for any board if bl_update is included.

**Comment by dagar** [comment]
There's a shellcheck failure in CI.

**Comment by dagar** [comment]
@PX4/testflights could you try this on a few px4fmu-v2 and px4fmu-v3 units?

 1. connect to board and run `ver all` over nsh (save output)
 2. flash through QGC (note the reported bootloader version)
 3. set the parameter SYS_BL_UPDATE 1 and reboot
 4. connect to board and run `ver all` over nsh (save output)
 5. flash through QGC again (note the reported bootloader version)

**Comment by bkueng** [comment]
@hamishwillee this one will require docs too. Can I leave that to you?

**Comment by hamishwillee** [comment]
> @hamishwillee this one will require docs too. Can I leave that to you?

@bkueng Sure, a few questions:
1. How will a user know that they need to do this? i.e. I'm a user with a board, I go to QGC and try download firmware but my bootloader is v2. What tells me I need to try update? What will I see in the QGC window? 
2. Is this step needed: *Insert an SD card (for the bootlog)*

I propose to add information on this as a separate section in https://docs.px4.io/en/config/firmware.html

3. Does that make sense to you?

**Comment by bkueng** [comment]
> How will a user know that they need to do this? i.e. I'm a user with a board, I go to QGC and try download firmware but my bootloader is v2. What tells me I need to try update? What will I see in the QGC window?

Basically whenever they run into the limited FLASH issue. I know this is not a good answer, but with this implementation QGC won't inform you that you need this. Which is why I think it would still be beneficial to have it integrated in QGC (https://github.com/mavlink/qgroundcontrol/issues/6488).
This is really only a minimal implementation so that we can point users to an easy way to update the BL.

> Is this step needed: Insert an SD card (for the bootlog)

Not strictly, but desired, so that we have a log when something goes wrong.

> I propose to add information on this as a separate section in https://docs.px4.io/en/config/firmware.html

That makes sense. Maybe we should also go through the places where we mention the limited v2 FLASH and add a link.

**Comment by santiago3dr** [comment]
looks to be working on the pixhawk mini
flashed stable through qgc; auto-detected as v2


flashed this pr, set parameter to update bootloader and reset
flashed stable through qgc; now auto-detected as v3

**Comment by santiago3dr** [comment]
PROBLEM!

grabbed an old pixhawk known to have a rev1 cpu with silicon errata
silicon errata

updated the bootloader and it's pulling v3 despite memory constraint
old pixhawk

**Comment by LorenzMeier** [comment]
@davids5 Sounds like the bootloader needs to be more clever about reporting flash size?

**Comment by davids5** [comment]
@LorenzMeier - hmm remember the pr for the booloader that just built V3 as V2? I suspect the flash size check is ifdefed to FMU_V2. So the Bootloader needs to be reved.

File: ROMFS/px4fmu_common/init.d/rcS [hunk 1]
```diff
@@ -211,6 +211,27 @@ else
 		param reset
 	fi
 
+	#
+	# Bootloader upgrade
+	#
+	if ver hwcmp PX4FMU_V2
+	then
+		set BL_FILE /etc/extras/px4fmuv3_bl.bin
```
Comment (davids5) [inline ROMFS/px4fmu_common/init.d/rcS]: @LorenzMeier as an alternative use  /etc/extras/px4fmuv2_bl.bin

**Comment by bkueng** [comment]
I updated the binary to https://github.com/PX4/Bootloader/pull/115.

@santiago3dr Thanks for testing. Can you go through the steps again with the updated branch?

**Comment by santiago3dr** [comment]
@bkueng looks to be working as expected

old pixhawk with silicon errata properly detects 1MB and pulls v2
old pixhawk

pixhawk mini with full 2MB pulls v3
pixhawk mini

**Comment by bkueng** [comment]
Great. Thanks for the support everyone.

**Pull request merged**